test/integration/validation_test.go

```diff
@@ -13,8 +13,233 @@ import (
 
 	"github.com/eggsampler/acme/v3"
 	"github.com/letsencrypt/boulder/test/vars"
+	"github.com/miekg/dns"
 )
 
+func TestMPICTLSALPN01(t *testing.T) {
+	t.Parallel()
+
+	client, err := makeClient()
+	if err != nil {
+		t.Fatalf("creating acme client: %s", err)
+	}
+
+	domain := randomDomain(t)
+
+	order, err := client.Client.NewOrder(client.Account, []acme.Identifier{{Type: "dns", Value: domain}})
+	if err != nil {
+		t.Fatalf("creating order: %s", err)
+	}
+
+	authz, err := client.Client.FetchAuthorization(client.Account, order.Authorizations[0])
+	if err != nil {
+		t.Fatalf("fetching authorization: %s", err)
+	}
+
+	chal, ok := authz.ChallengeMap[acme.ChallengeTypeTLSALPN01]
+	if !ok {
+		t.Fatalf("no TLS-ALPN-01 challenge found in %#v", authz)
+	}
+
+	_, err = testSrvClient.AddARecord(domain, []string{"10.88.88.88"})
+	if err != nil {
+		t.Fatalf("adding A record: %s", err)
+	}
+	defer func() {
+		testSrvClient.RemoveARecord(domain)
+	}()
+
+	_, err = testSrvClient.AddTLSALPN01Response(domain, chal.KeyAuthorization)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		_, err = testSrvClient.RemoveTLSALPN01Response(domain)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	chal, err = client.Client.UpdateChallenge(client.Account, chal)
+	if err != nil {
+		t.Fatalf("completing TLS-ALPN-01 validation: %s", err)
+	}
+
+	validationEvents, err := testSrvClient.TLSALPN01RequestHistory(domain)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	validationCount := 0
+	for _, event := range validationEvents {
+		if event.ServerName == domain {
+			validationCount++
+		}
+	}
+	if validationCount != 4 {
+		t.Fatalf("expected 4 validation events (VA=1 RVAs=3), got %d", validationCount)
```

Comment:
This should be an `Errorf` because the test can continue and collect more data (like whether the DNS request history is correct).